### ISSUE TITLE
IllegalArgumentException when reading from a freshly-created queue containg just one message

### ISSUE THREAD

**Issue by nanomad** [OpenHFT/Chronicle-Queue#619]
Hi, we are encountering a `java.lang.IllegalArgumentException: Unsigned Int 31-bit -1 out of range` exception after reading from a queue in the following state:

- Just one message in the queue
- The read happens an interval of time after the write that is less than the rolling cycle of the queue

Tested version: chronicle-queue 5.17.23
 
MWE / Test case is after the stack trace.

Stack trace
```
java.lang.IllegalArgumentException: Unsigned Int 31-bit -1 out of range

	at net.openhft.chronicle.core.Maths.toUInt31(Maths.java:373)
	at net.openhft.chronicle.queue.RollCycles.toCycle(RollCycles.java:105)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndexResult(SingleChronicleQueueExcerpts.java:1408)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndexInternal(SingleChronicleQueueExcerpts.java:1467)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndex(SingleChronicleQueueExcerpts.java:1404)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.exceptsPerCycle(SingleChronicleQueue.java:490)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.tryWindBack(SingleChronicleQueueExcerpts.java:1769)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.windBackCycle(SingleChronicleQueueExcerpts.java:1757)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.incrementIndex(SingleChronicleQueueExcerpts.java:1745)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer$StoreTailerContext.close(SingleChronicleQueueExcerpts.java:2017)
	at net.openhft.chronicle.wire.MarshallableIn.readText(MarshallableIn.java:105)
	at it.coderit.samples.cqbug.CqBug.testReadAfterWriteJustOneMessage(CqBug.java:55)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
java.lang.IllegalArgumentException: Unsigned Int 31-bit -1 out of range

	at net.openhft.chronicle.core.Maths.toUInt31(Maths.java:373)
	at net.openhft.chronicle.queue.RollCycles.toCycle(RollCycles.java:105)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndexResult(SingleChronicleQueueExcerpts.java:1408)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndexInternal(SingleChronicleQueueExcerpts.java:1467)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndex(SingleChronicleQueueExcerpts.java:1404)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.exceptsPerCycle(SingleChronicleQueue.java:490)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.tryWindBack(SingleChronicleQueueExcerpts.java:1769)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.windBackCycle(SingleChronicleQueueExcerpts.java:1757)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.incrementIndex(SingleChronicleQueueExcerpts.java:1745)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer$StoreTailerContext.close(SingleChronicleQueueExcerpts.java:2017)
	at net.openhft.chronicle.wire.MarshallableIn.readText(MarshallableIn.java:105)
	at it.coderit.samples.cqbug.CqBug.testReadAfterWriteJustOneMessage(CqBug.java:55)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

MWE / Test case:

```java
package it.coderit.samples.cqbug;

import net.openhft.chronicle.core.time.SetTimeProvider;
import net.openhft.chronicle.queue.ChronicleQueue;
import net.openhft.chronicle.queue.ExcerptTailer;
import net.openhft.chronicle.queue.RollCycles;
import net.openhft.chronicle.queue.TailerDirection;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;
import java.util.Date;
import java.util.concurrent.TimeUnit;

import static junit.framework.TestCase.assertEquals;

public class CqBug {


    @Rule
    public TemporaryFolder temporaryFolder = new TemporaryFolder();

    private File dataDir;
    private SetTimeProvider timeProvider;

    @Before
    public void setup() throws IOException {
        this.dataDir = temporaryFolder.newFolder();
        this.timeProvider = new SetTimeProvider(new Date().getTime());
    }

    @Test
    public void testReadAfterWriteJustOneMessage() {
        RollCycles rollingCycle = RollCycles.DAILY;
        // Write a message to the queue
        try (ChronicleQueue queue = ChronicleQueue.singleBuilder(dataDir)
                .timeProvider(timeProvider)
                .rollCycle(rollingCycle)
                .build()) {
            queue.acquireAppender().writeText("42");
        }

        // Wait less than the rolling cycle
        timeProvider.advanceMillis(TimeUnit.HOURS.toMillis(6));

        // Read backwards from the end
        try (ChronicleQueue queue = ChronicleQueue.singleBuilder(dataDir)
                .timeProvider(timeProvider)
                .rollCycle(rollingCycle)
                .build()) {
            ExcerptTailer tailer = queue.createTailer().direction(TailerDirection.BACKWARD).toEnd();
            // An exception is thrown here
            String read = tailer.readText();
            assertEquals("42", read);
        }

    }

}
```

**Comment by dpisklov** [comment]
@nanomad thanks for reporting the issue and especially for the test case, I have now fixed it.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.24](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.24), [BOM-2.17.396](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.396)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.24](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.24), [BOM-2.17.407](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.407)
